lib/sdk/tabs/tab-firefox.js

```diff
@@ -3,309 +3,328 @@
-   */
-  get id() this._tab ? getTabId(this._tab) : undefined,
-
-  /**
-   * The title of the page currently loaded in the tab.
-   * Changing this property changes an actual title.
-   * @type {String}
-   */
-  get title() this._tab ? getTabTitle(this._tab) : undefined,
-  set title(title) this._tab && setTabTitle(this._tab, title),
-
-  /**
-   * Returns the MIME type that the document loaded in the tab is being
-   * rendered as.
-   * @type {String}
-   */
-  get contentType() this._tab ? getTabContentType(this._tab) : undefined,
-
-  /**
-   * Location of the page currently loaded in this tab.
-   * Changing this property will loads page under under the specified location.
-   * @type {String}
-   */
-  get url() this._tab ? getTabURL(this._tab) : undefined,
-  set url(url) this._tab && setTabURL(this._tab, url),
-  /**
-   * URI of the favicon for the page currently loaded in this tab.
-   * @type {String}
-   */
   get favicon() {
     deprecateUsage(
       'tab.favicon is deprecated, ' +
       'please use require("sdk/places/favicon").getFavicon instead.'
     );
-    return this._tab ? getFaviconURIForLocation(this.url) : undefined
+    return isDestroyed(this) ? undefined : getFaviconURIForLocation(this.url);
   },
-  /**
-   * The CSS style for the tab
-   */
-  get style() null, // TODO
-  /**
-   * The index of the tab relative to other tabs in the application window.
-   * Changing this property will change order of the actual position of the tab.
-   * @type {Number}
-   */
-  get index()
-    this._tab ?
-    this._window.gBrowser.getBrowserIndexForDocument(this._contentDocument) :
-    undefined,
-  set index(value)
-    this._tab && this._window.gBrowser.moveTabTo(this._tab, value),
-  /**
-   * Thumbnail data URI of the page currently loaded in this tab.
-   * @type {String}
-   */
-  getThumbnail() {
-    if (!this._tab)
-      return undefined;
-    if (this._tab.getAttribute('remote')) {
-      console.error('This method is not supported with E10S');
-      return BLANK;
-    }
-    return getThumbnailURIForWindow(this._contentWindow);
+
+  get contentType() {
+    return isDestroyed(this) ? undefined : browser(this).documentContentType;
   },
-  /**
-   * Whether or not tab is pinned (Is an app-tab).
-   * @type {Boolean}
-   */
-  get isPinned() this._tab ? this._tab.pinned : undefined,
-  pin: function pin() {
-    if (!this._tab)
-      return;
-    this._window.gBrowser.pinTab(this._tab);
+
+  get index() {
+    return isDestroyed(this) ? undefined : getIndex(viewsFor.get(this));
   },
-  unpin: function unpin() {
-    if (!this._tab)
+
+  set index(val) {
+    if (isDestroyed(this))
       return;
-    this._window.gBrowser.unpinTab(this._tab);
+
+    move(viewsFor.get(this), val);
+  },
+
+  get isPinned() {
+    return isDestroyed(this) ? undefined : viewsFor.get(this).pinned;
+  },
+
+  get window() {
+    if (isClosed(this))
+      return undefined;
+
+    require('../windows');
```

Comment:
please add a `// TODO: Remove the dependency on the windows module: bug 792670` comment here too